File: Progs/Sorting/QuickSort.java
public class QuickSort {
    public static void main(String[] args) {
	new QuickSort().start(); 
    }
    
    public void start() {
	int arr[] = {23, 675, 1 ,6, 12, 9, 1, 435, 6, 567};
	quickSort(arr, 0, 9);
	display(arr);
    }

    public void quickSort(int[] arr, int lower_index, int upper_index) {
	if(lower_index < upper_index) {
	    int index = placeInRightPosition(arr, lower_index, upper_index);
	    quickSort(arr, lower_index, index - 1);
	    quickSort(arr, index + 1, upper_index);
	}
    }

    public int placeInRightPosition(int[] arr, int pivot_index, int upper_index) {
	int lower_index = pivot_index + 1;
	int org_upper_index = upper_index;
	while (lower_index < upper_index) {
	    while (lower_index < org_upper_index && arr[pivot_index] >= arr[lower_index]) {
		lower_index++;
	    }
	    while (upper_index > (pivot_index + 1) && arr[pivot_index] <= arr[upper_index]) {
		upper_index--;
	    }
	    if (lower_index < upper_index) {
		int t = arr[lower_index];
		arr[lower_index] = arr[upper_index];
		arr[upper_index] = t;
	    }
	}
	if (arr[upper_index] < arr[pivot_index]) {
	    int t = arr[upper_index];
	    arr[upper_index] = arr[pivot_index];
	    arr[pivot_index] = t;
	    pivot_index = upper_index;
	}
	return pivot_index;
    }

    public void display(int[] arr) {
	for (int i = 0; i < arr.length; i++) {
	    System.out.println(arr[i]);
	}
    }
}

File: Progs/Practice/Sorting/QuickSort.cpp
#include <iostream>

using namespace std;

void display(int *arr)
{int i;
 for(i=0; i<10; i++)
	cout << arr[i] << "\t";
 cout << endl;
}

void swap(int *a, int *b)
{
 int temp=*a;
 *a=*b;
 *b=temp;
}

void Quick_Sort(int *arr, int start, int end)
{
 int istart = start;
 int iend = end;
 int pivot = start++;
 
 display(arr);
 
 while(start<end)
	{
	 while(arr[start] < arr[pivot] && start < iend)
		start++;
	 while(arr[end] > arr[pivot] && end>(pivot+1))
		end--;
	 if(start<end)
		swap(&arr[start], &arr[end]);
	}
 if(arr[pivot] > arr[end])
 	swap(&arr[pivot], &arr[end]);
 cout << pivot << " " << end << " " << start << " " << istart << " " << iend << endl;
 if(end-1-istart>0)
	Quick_Sort(arr, istart, end-1);
 if(iend-end-1>0)
	Quick_Sort(arr, end+1, iend);
}

int main()
{
 int arr[]={23,54,2,45,65,23,5,68,43,34}, i;
 Quick_Sort(arr, 0, 9);
 display(arr);
}

